pol-core/bscript/compiler/astbuilder/FunctionResolver.cpp

```diff
@@ -85,55 +113,97 @@ void FunctionResolver::register_module_function( ModuleFunctionDeclaration* mf )
+  if ( uf->type == UserFunctionType::Constructor )
+  {
+    resolved_functions[{ ScopeName::Global, name }] = uf;
+  }
 }
 
 bool FunctionResolver::resolve( std::vector<AvailableParseTree>& to_build_ast )
 {
-  for ( auto unresolved_itr = unresolved_function_links_by_name.begin();
-        unresolved_itr != unresolved_function_links_by_name.end(); )
+  // Attempt to link all unresolved function links
+  for ( auto unresolved_itr = unresolved_function_links.begin();
+        unresolved_itr != unresolved_function_links.end(); )
   {
-    const std::string& name = ( *unresolved_itr ).first;
-    const std::vector<std::shared_ptr<FunctionLink>>& function_links = ( *unresolved_itr ).second;
-
-    auto previously_resolved_itr = resolved_functions_by_name.find( name );
-    if ( previously_resolved_itr != resolved_functions_by_name.end() )
+    for ( auto function_link_itr = ( *unresolved_itr ).second.begin();
+          function_link_itr != ( *unresolved_itr ).second.end(); )
     {
-      Function* resolved_function = ( *previously_resolved_itr ).second;
-      for ( auto& function_link : function_links )
+      auto& function_link = *function_link_itr;
+      const auto& calling_scope = function_link->calling_scope;
+      const auto& name = ( *unresolved_itr ).first;
+      const auto& unscoped_name = name.name;
+      const auto& call_scope = name.scope;
+
+      // Given a Function, link the FunctionLink to it and remove it from the list.
+      // If the function is already resolved, link it.
+      auto resolve_if_existing = [&]( const ScopableName& key )
       {
-        function_link->link_to( resolved_function );
-      }
+        auto resolved_function = check_existing( key );
+        if ( resolved_function )
+        {
+          ( *function_link_itr )->link_to( resolved_function );
+          function_link_itr = ( *unresolved_itr ).second.erase( function_link_itr );
+          report.debug( function_link->source_location, "linking ({}, {}) to {}::{}",
+                        call_scope.string(), name.string(), resolved_function->module_name,
+                        resolved_function->name );
+          return true;
+        }
+        return false;
+      };
+
+      auto advance_if_build_available = [&]( const ScopableName& key )
+      {
+        if ( build_if_available( to_build_ast, calling_scope, key ) )
+        {
+          ++function_link_itr;
+          return true;
+        }
+        return false;
+      };
+
+      // Link the function if possible, otherwise try to build it.
+      auto link_handled = [&]( const ScopableName& key )
+      { return resolve_if_existing( key ) || advance_if_build_available( key ); };
+
+      report.debug( function_link->source_location, "resolving {}", name.string() );
+
+      if ( !call_scope.empty() && link_handled( { call_scope, unscoped_name } ) )
+        continue;
+
+      // If calling scope is present, check that first, eg. nside Animal class, `foo()` checks
```

Comment:
again, incorrect comment. furthermore, can this functionality be modularized, as `resolve_if_existing` does the same path